docs-source/content/guides/configuration-file.md
---
title: Configuration file
type: docs
---

# Configuration file

Each Hereditas working folder contains a JSON configuration file called `hereditas.json`. This file is automatically generated when running the [`hereditas init`]({{< relref "/cli/init.md" >}}) command, and it's modified by the app itself under certain circumstances.

## Contents

The structure of the document is similar to the following:

````json
{
    "version": 20190222,
    "contentDir": "content",
    "distDir": "dist",
    "appToken": "...",
    "waitTime": 86400,
    "users": [
        {
            "email": "me@example.com",
            "role": "owner"
        },
        {
            "email": "someone@example.com",
            "role": "user"
        }
    ],
    "auth0": {
        "domain": "myhereditas.auth0.com",
        "managementClientId": "...",
        "managementClientSecret": "...",
        "hereditasClientId": "...",
        "rules": [
            '...',
            '...',
            '...'
        ]
    },
    "urls": [
        "https://my.testhereditas.app",
        "https://another.testhereditas.app"
    ],
    "webhookUrl": "https://example.com/webhook/token/abc123XYZ",
    "kdf": "pbkdf2",
    "pbkdf2": {
        "iterations": 100000
    },
    "processMarkdown": true
}
````

## Configuration options

### Hereditas working folder options

These options configure the way the Hereditas current working folder is set up.

- **`version`** (int): This represents the version of the configuration file. At present moment, this is `20190222`.
- **`contentDir`** (string): Name of the directory, inside the Hereditas working folder, containing the documents and files to include in the Hereditas box. The content of this folder is not encrypted, as the CLI will do that automatically.<br />The default value is `content`. This value can be set during box initialization with the `--content` option for the [`hereditas init`]({{< relref "/cli/init.md" >}}) command.
- **`distDir`** (string): Name of the directory, inside the Hereditas working folder, where the CLI will put generated web app and the encrypted content.<br />The default value is `dist`. This value can be set during box initialization with the `--dist` option for the [`hereditas init`]({{< relref "/cli/init.md" >}}) command.

### Access control basics

These options are used to configure the way Hereditas protects access to your box.

- **`appToken`** (string): The *application token*, which together with the *user passphrased* is used to derive the encryption and decryption key. The application token is stored in the Hereditas configuration file and synced with Auth0. This is an important secret to protect, although in itself (without the user passphrase, which isn't stored anywhere) isn't sufficient to encrypt or decrypt your data.<br/>This value is generated automatically when a new Hereditas project is generated (using [`hereditas init`]({{< relref "/cli/init.md" >}})), and can be re-generated with [`hereditas regenerate-token`]({{< relref "/cli/regenerate-token.md" >}}).
- **`waitTime`** (string): Amount of time users need to wait before they can unlock an Hereditas box, in seconds. The default value is 86400, or 24 hours.<br/>This value can be also set with [`hereditas wait-time:set`]({{< relref "/cli/wait-time_set.md" >}})
- **`users`** (array of objects): Array of users that are whitelisted to use the app. Users can authenticate with Auth0 using any social account they support, e.g. Google, Facebook, Microsoft, etc; as long as the email address matches what's whitelisted in this configuration option. For more information, please refer to the documentation on [Managing users]({{< relref "/guides/managing-users.md" >}}). This value is an array of objects with the structure:
    - **`users.$.email`** (string): email address of the user
    - **`users.$.role`** (string): either `user` or `owner`

### Auth0 settings and credentials

The **`auth0`** object contains settings and credentials for communicating with the Auth0 APIs. For more information, please refer to the [Auth0 Setup]({{< relref "/guides/auth0-setup.md" >}}) article.

- **`auth0.domain`** (string): domain on Auth0, e.g. `myhereditas.auth0.com`. This is created when you sign up for Auth0, and it's globally unique.<br/>This is set with the `--auth0Domain` option for the [`hereditas init`]({{< relref "/cli/init.md" >}}) command.
- **`auth0.managementClientId`** (string): Client Id (ie. public key) for the "API Access" application on Auth0, a "Machine-to-Machine" application used by the Hereditas CLI and rules to interact with the Auth0 APIs.<br/>This is set with the `--auth0ClientId` option for the `hereditas init` command.
- **`auth0.managementClientSecret`** (string): Client Secret (ie. private key) for the "API Access" application on Auth0.<br/>This is set with the `--auth0ClientSecret` option for the `hereditas init` command.
- **`auth0.hereditasClientId`** (string): Client Id (ie. public key) for the "Hereditas" application on Auth0. This value is generated automatically by the Auth0 CLI, when running the [`hereditas auth0_sync`]({{< relref "/cli/auth0_sync.md" >}}) command. In most cases, you should not edit this Client Id manually.
- **`auth0.rules`** (array of strings): These strings represent the Ids of the rules that the Hereditas CLI creates on Auth0. These are returned automatically by the Auth0 CLI, when running the [`hereditas auth0_sync`]({{< relref "/cli/auth0_sync.md" >}}) command. In most cases, you should not edit this configuration setting manually.

### Deployment and webhook URLs

The Hereditas configuration file stores two separate kinds of URLs:

- **`urls`** (array of strings): list of URLs where the Hereditas application is deployed to. These URLs are synced with Auth0, which uses them to whitelist callback URLs after users authenticate successfully. You should list every URL where your app might be reachable at. Note also that the protocol needs to match too, so `http://example.com` and `https://example.com` are separate URLs. For more information, see the [Auth0 Setup]({{< relref "/guides/auth0-setup.md" >}}) and [Deploying the box]({{< relref "/guides/deploy-box.md" >}}) articles.<br />You need to provide (at least) one URL with the `--url` flag when running [`hereditas init`]({{< relref "/cli/init.md" >}}). These URLs can also be changed with CLI commands, such as [`url:add`]({{< relref "/cli/url_add.md" >}}) and [`url:rm`]({{< relref "/cli/url_rm.md" >}})].
- **`webhookUrl`** (string): URL of the webhook that is invoked to notify owners of a successful authentication. Please refer to the [Login notifications]({{< relref "/guides/login-notifications.md" >}}) article for more information.<br />This value can also be set with the [`webhook:set`]({{< relref "/cli/webhook_set.md" >}}) command.

### Advanced options

These options are set by default by Hereditas. You shouldn't change these options unless you have a good reason for it and you're confident that you know what you're doing.

- **`kdf`** (string): Key derivation function to use. Currently, only `pbkdf2` is supported.
- **`pbkdf2`** (object): Parameters for pbkdf2.
    - **`pbkdf2.iterations`** (int): Number of iterations to use for pbkdf2. Default value is 100000 (1E+06)
- **`processMarkdown`** (boolean): Switch to enable/disable the conversion of Markdown documents into HTML, when building the box. Default is true (enabled).


docs-source/content/advanced/Configuration-file.md
---
title: Configuration file
type: docs
---

# Configuration file


cli/lib/Config.js
'use strict'

const fs = require('fs')
const util = require('util')
const defaultsDeep = require('lodash.defaultsdeep')
const cloneDeep = require('lodash.clonedeep')
const SMHelper = require('smhelper')

const ConfigVersion = 20190222

/**
 * Authorized users
 *
 * @typedef {object} HereditasUser
 * @property {string} email - Email address
 * @property {"user"|"owner"} role - Role: "user" or "owner"
 */

/**
 * Configuration dictionary for Hereditas
 *
 * @typedef {object} HereditasConfig
 * @property {number} version - Version of the configuration object
 * @property {string} contentDir - Folder containing the source content
 * @property {string} distDir - Folder where to place the compiled project
 * @property {boolean} processMarkdown - If true, enable processing of Markdown files into HTML
 * @property {object} auth0 - Auth0 configuration
 * @property {string} auth0.domain - Auth0 domain/tenant (e.g. "myhereditas.auth0.com")
 * @property {string} auth0.hereditasClientId - Auth0 app client ID for Hereditas
 * @property {string} auth0.managementClientId - Client ID for the Auth0 Management app
 * @property {string} auth0.managementClientSecret - Client Secret for the Auth0 Management app
 * @property {Array<string>} rules - ID of the Auth0 rules created by the Hereditas CLI
 * @property {"pbkdf2"|"argon2"} kdf - Key derivation function to use: pbkdf2 (default) or argon2
 * @property {object} pbkdf2 - Configuration for pbkdf2
 * @property {string} pbkdf2.iterations - Number of iterations
 * @property {string} webhookUrl - URL of the webhook to trigger when a new user logs into Hereditas.
 * @property {Array<HereditasUser>} users - List of users
 * @property {string} appToken - Application token; when combined with the user passphrase, this allows deriving the encryption key
 * @property {number} waitTime - The amount of time, in seconds, to wait before Auth0 can return to users the app token
 * @property {Array<string>} urls - list of URLs where your app will be deployed to, e.g. `https://hereditas.example.com`, or `https://myname.blob.core.windows.net/hereditas`, etc; this is used for OAuth redirects.
 */

/**
 * Helper class for managing Hereditas configuration
 */
class Config {
    /**
     * Initializes the object.
     *
     * @param {string} [filename="hereditas.json"] - Name of the file on disk
     */
    constructor(filename) {
        if (!filename) {
            filename = 'hereditas.json'
        }

        this._filename = filename

        // userConfig is the data read from the config file. config is that, plus defaults
        this._userConfig = null
        this._config = {}
    }

    /**
     * Create a new userConfig object.
     *
     * Note that this doesn't save changes on disk, you must manually call `save()`.
     *
     * @param {HereditasConfig} initConfig - Initial configuration values
     */
    create(initConfig) {
        this._userConfig = {
            version: ConfigVersion
        }
        defaultsDeep(this._userConfig, initConfig)

        // Update the config in memory
        this._config = {}
        this._defaults()
    }

    /**
     * Reads and parses a config file, validating it.
     *
     * @param {string} [filename="hereditas.json"] - Name of the config file to read; default is "hereditas.json"
     * @throws Throws an error if the config file doesn't exist or is not a valid Hereditas config
     */
    async load() {
        // Read the file
        const configFile = await util.promisify(fs.readFile)(this._filename, 'utf8')
        if (!configFile) {
            throw Error('Cannot read config file')
        }

        // Parse JSON and ensure it's a valid format
        this._userConfig = JSON.parse(configFile)
        if (!this._validate()) {
            throw Error('Invalid config file')
        }

        // Apply defaults
        this._defaults()
    }

    /**
     * Returns value for key from configuration
     *
     * @param {string} key - Key of the object, in "dot notation"
     * @returns {*} Value of the configuration key requested (cloned)
     */
    get(key) {
        let val

        // If key contains a dot, we are requesting a nested object
        if (key.indexOf('.') != -1) {
            val = SMHelper.getDescendantProperty(this._config, key)
        }
        else {
            val = this._config[key]
        }

        // Returns a clone of the object so it can't be modified
        return cloneDeep(val)
    }

    /**
     * Returns all config values (cloned).
     *
     * @returns {HereditasConfig} All configuration data
     */
    all() {
        return cloneDeep(this._config)
    }

    /**
     * Updates the value of a user config.
     *
     * Note: this does NOT save the changes on disk; you must invoke `save()` for that.
     *
     * @param {string} key - Name of the key to update, using the "dot notation"
     * @param {*} value - New value
     */
    set(key, value) {
        // Update the value and validate the config
        SMHelper.updatePropertyInObject(this._userConfig, key, value)
        if (!this._validate()) {
            throw Error('Invalid config data')
        }

        // Update the config in memory
        this._config = {}
        this._defaults()
    }

    /**
     * Save changes to user configuration to disk.
     *
     * @returns {Promise} Returns a promise that resolves when the changes have been committed to disk.
     * @async
     */
    save() {
        return util.promisify(fs.writeFile)(this._filename, JSON.stringify(this._userConfig, null, 2))
    }

    /**
     * Validates a config object, ensuring that all required keys are present.
     *
     * @returns {boolean} Returns true on valid configuration objects
     * @throws Throws an Error if the config file isn't valid
     */
    _validate() {
        if (!this._userConfig || typeof this._userConfig != 'object' || !Object.keys(this._userConfig).length) {
            throw Error('Invalid config file')
        }
        if (!this._userConfig.version) {
            throw Error('Config file is missing required key version')
        }
        if (!this._userConfig.distDir) {
            throw Error('Config file is missing required key distDir')
        }
        if (!this._userConfig.contentDir) {
            throw Error('Config file is missing required key contentDir')
        }
        if (!this._userConfig.appToken) {
            throw Error('Config file is missing required key appToken')
        }
        if (!this._userConfig.auth0 || typeof this._userConfig.auth0 != 'object' || !Object.keys(this._userConfig.auth0).length) {
            throw Error('Config file is missing required key auth0')
        }
        if (!this._userConfig.auth0.domain) {
            throw Error('Config file is missing required key auth0.domain')
        }
        if (!this._userConfig.urls || !Array.isArray(this._userConfig.urls) || !this._userConfig.urls.length) {
            throw Error('Config file is missing required key urls')
        }

        return true
    }

    /**
     * Applies default parameters to the userConfig object, and stores that into the config object
     */
    _defaults() {
        defaultsDeep(this._config, this._userConfig, {
            processMarkdown: true,
            kdf: 'pbkdf2',
            pbkdf2: {
                iterations: 100000
            }
        })
    }
}

module.exports = Config
